removed usage of numpy's private functions from util.arraycrop

Harry-Kwon opened the pull request:
## Description

Modified skimage.util.arraycrop to remove references to numpy's private method _as_pair.
Also added tests to more closely match the documentation of the method.
Fixed documentation to match parameter format already implemented in method and in tests.

Closes issue #5028 
I did not find any other places where numpy's private methods were being used.

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @Harry-Kwon! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/util/arraycrop.py`](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/arraycrop.py):

> [Line 23:80](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/arraycrop.py#L23): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 54:25](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/arraycrop.py#L54): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 65:19](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/arraycrop.py#L65): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`skimage/util/tests/test_arraycrop.py`](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/tests/test_arraycrop.py):

> [Line 22:1](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/tests/test_arraycrop.py#L22): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 36:1](https://github.com/scikit-image/scikit-image/blob/faaf9f3ace955f0c78b3844b1b0474efc37ae2bf/skimage/util/tests/test_arraycrop.py#L36): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

jjerphan commented on skimage/util/arraycrop.py:
```diff
@@ -4,24 +4,23 @@
 
-def crop(ar, crop_width, copy=False, order='K'):
-    """Crop array `ar` by `crop_width` along each dimension.
+def crop(array, crop_width, copy=False, order='K'):
```
This is a proper naming, but this will break the API.

jjerphan replied:
This is a proper renaming, but this will break the API. Reverting this change is safer.

jjerphan commented on skimage/util/arraycrop.py:
```diff
@@ -38,13 +37,28 @@ def crop(ar, crop_width, copy=False, order='K'):
+        elif len(crop_width) == 2:
+            crops = [crop_width] * array.ndim
+        else:
+            raise ValueError('invalid length for sequence crop_width')
```
Maybe a message giving the length of `crop_width` and the expected ones will help?

jjerphan commented on skimage/util/arraycrop.py:
```diff
@@ -38,13 +37,28 @@ def crop(ar, crop_width, copy=False, order='K'):
+    elif len(crop_width) == array.ndim:
+        crops = crop_width
+    else:
+        raise ValueError('invalid length for sequence crop_width')
```
Ditto.

jjerphan reviewed (changes requested):
Hi @Harry-Kwon,

Apart the few suggestions left in your PRs, everything LGTM.

Harry-Kwon commented:
> Hi @Harry-Kwon,
> 
> Apart the few suggestions left in your PRs, everything LGTM.

Thanks for the feedback @jjerphan!

I've added a more descriptive error message explaining expected input values as you suggested. I'm not sure if the phrase `a sequence of N pairs` is too ambiguous on its own but I think it makes sense in the context of the docstring.

I've also reverted name change you suggested, but I am curious as to where it breaks the API. If the name `array` is preferable to `ar` where would I need to make a change to make the renaming not break the api?

jjerphan commented on skimage/util/arraycrop.py:
```diff
@@ -37,28 +37,36 @@ def crop(array, crop_width, copy=False, order='K'):
         else:
-            raise ValueError('invalid length for sequence crop_width')
+            raise ValueError(
+                f"crop_width sequence of length {len(crop_width)} is invalid\n"
```
```
crop_width has a invalid length: {len(crop_width)}.
```
is a simpler phrasing.

Harry-Kwon replied:
I've implemented this in a new commit.

jjerphan commented on skimage/util/arraycrop.py:
```diff
@@ -37,28 +37,36 @@ def crop(array, crop_width, copy=False, order='K'):
     else:
-        raise ValueError('invalid length for sequence crop_width')
+        raise ValueError(
+            f"crop_width sequence of length {len(crop_width)} is invalid\n"
```
Ditto.

jjerphan commented:
> Thanks for the feedback @jjerphan!

You're welcome!

> I've added a more descriptive error message explaining expected input values as you suggested. I'm not sure if the phrase a sequence of N pairs is too ambiguous on its own but I think it makes sense in the context of the docstring.

It makes sense to me.

> I've also reverted name change you suggested, but I am curious as to where it breaks the API. If the name array is preferable to ar where would I need to make a change to make the renaming not break the api?

If someone previously have used it like:

```python
a = crop(arr=…, crop_width=…)
```

Changing `arr` to `array` will break their code.

Harry-Kwon commented:
> If someone previously have used it like:
> 
> ```python
> a = crop(arr=…, crop_width=…)
> ```
> 
> Changing `arr` to `array` will break their code.

Oh right, I see. I guess it wouldn't be possible to rename it now without breaking the api.

jni reviewed (approved):
Awesome, thank you @Harry-Kwon for the PR and @jjerphan for the review!

alexdesiqueira commented:
Thank you @Harry-Kwon @jjerphan!
